source/common/upstream/load_balancer_impl.cc

```diff
@@ -828,6 +858,21 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
   return candidate_host;
 }
 
+HostConstSharedPtr RandomLoadBalancer::peekAnotherHost(LoadBalancerContext* context) {
+  const absl::optional<HostsSource> hosts_source = hostSourceToUse(context);
+  if (!hosts_source) {
+    return nullptr;
+  }
+
+  const HostVector& hosts_to_use = hostSourceToHosts(*hosts_source);
+  if (hosts_to_use.empty()) {
+    return nullptr;
+  }
+
+  stashed_random_.push_back(random_.random());
+  return hosts_to_use[stashed_random_.back() % hosts_to_use.size()];
+}
+
 HostConstSharedPtr RandomLoadBalancer::chooseHostOnce(LoadBalancerContext* context) {
```

Comment:
Thoughts on a possibly more generic implementation that would apply to Random, WRR and LeastLoaded LBs:

Move the current implementations of chooseHostOnce to chooseNextHostInternal.

peekAnotherHost would call chooseNextHostInternal to add 1 entry to a std::vector<weak_ptr<Host>> prefetch_hosts_ and return HostConstSharedPtr to the added Host.

chooseHostOnce would return the first non-expired entry in prefetch_hosts_ if any, otherwise return chooseNextHostInternal

Reply:
I don't think least loaded is safe to prefetch for.  My concern is multiple threads would all select the least loaded endpoint, and all stick it in the prefetch queue, where using it live, one would pick it and immediately add load, and then the other threads would not pick it.

The reason I didn't do prefetch the way you suggest is that it turns out that picking isn't 100% deterministic even for those load balancers.  If you drill down from chooseHostOnce/peekAnotherHost -> hostSourceToUse -> chooseHostSet, it consults the session for pluggable logic (only on retries, only if configured).  In this case the peek won't match the pick.

I looked into if I could refactor chooseHost to literally pass a nullptr context if it was not going to do custom logic for determinePriorityLoad, but the context is used in several other places for other functions, so that's not plausible.

I could instead add a function for hasCustomPriorityLoad(), and only pre-pick and use prepicks if hasCustomPriorityLoad is false.  WDYT?

Reply:
Great points about least-loaded, I had forgotten that we need to consider multiple threads :)

I think I'm fine with either the current approach or the hasCustomPriorityLoad() == false option.  The former has the benefit of already being implemented.